modules/renter/repair_test.go

```diff
@@ -48,27 +98,48 @@ func TestRepair(t *testing.T) {
 		}
 	}
 
-	// file should not be available after first pass
-	if f.available() {
-		t.Fatalf("file should not be available: %v%%", f.uploadProgress())
+	// download data
+	chunks := make([][][]byte, f.numChunks())
+	for i := uint64(0); i < f.numChunks(); i++ {
+		chunks[i] = make([][]byte, rsc.NumPieces())
 	}
-
-	// repair until file becomes available
-	const maxAttempts = 20
-	for i := 0; i < maxAttempts; i++ {
-		for chunk, pieces := range f.incompleteChunks() {
-			err = f.repair(chunk, pieces, r, hosts)
+	for _, h := range hosts {
+		contract, exists := f.contracts[h.ContractID()]
+		if !exists {
+			continue
+		}
+		for _, p := range contract.Pieces {
+			encPiece := h.(*testHost).data[p.Offset : p.Offset+pieceSize+crypto.TwofishOverhead]
+			piece, err := deriveKey(f.masterKey, p.Chunk, p.Piece).DecryptBytes(encPiece)
 			if err != nil {
 				t.Fatal(err)
 			}
+			chunks[p.Chunk][p.Piece] = piece
 		}
-		if f.available() {
-			break
+	}
+	buf := new(bytes.Buffer)
+	for _, chunk := range chunks {
+		err = rsc.Recover(chunk, f.chunkSize(), buf)
+		if err != nil {
+			t.Fatal(err)
 		}
 	}
-	if !f.available() {
-		t.Fatalf("file not repaired to availability after %v attempts: %v", maxAttempts, err)
+	buf.Truncate(dataSize)
+
+	if !bytes.Equal(buf.Bytes(), data) {
+		t.Fatal("recovered data does not match original")
 	}
+
+	/*
+		for i, h := range hosts {
+			host := h.(*testHost)
+			pieces := 0
+			for _, p := range host.pieceMap {
+				pieces += len(p)
+			}
+			t.Logf("Host #: %d\tDelay: %v\t# Pieces: %v\t# Chunks: %d", i, host.delay, pieces, len(host.pieceMap))
+		}
+	*/
```

Comment:
I used it to assess how efficient the repair function is. Might be useful in the future, but if you leave it uncommented it will add lots of logging output to `make test`, which I didn't want.